src/lib.rs

```
@@ -3,11 +3,76 @@
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Pool {
+    /// Unique id of the mining pool. Can, for example, be used as a database id.
+    pub id: u64,
+    /// Name of the mining pool.
+    pub name: String,
+    /// The addresses the mining pool uses or used in the past.
+    pub addresses: Vec<String>,
+    /// The coinbase tags the mining pool uses or used in the past.
+    pub tags: Vec<String>,
+    /// Optional link to the mining pools website.
+    pub link: String,
+}
+
+impl Pool {
+    pub fn addresses(self, network: Network) -> Vec<Address> {
+        self.addresses
+            .iter()
+            .map(|a| {
+                let parsed = Address::from_str(a)
+                    .unwrap()
+                    .require_network(network)
+                    .unwrap();
+                // As the HRP 'tb1' is used for both testnet and signet, parsing
+                // a Signet address returns a testnet address in rust-bitcoin.
+                // We make sure to return a signet address in case the network
+                // is signet
+                if network == Network::Signet {
+                    Address::new(Network::Signet, parsed.payload().clone())
+                } else {
+                    parsed
+                }
+            })
+            .collect()
+    }
+}
+
+/// Parses a JSON formatted list of pools entries. Format:
+/// ```json
+/// [
+/// {
+///  "id": 7,
+///  "name": "Example Pool",
+///  "addresses": [
+///    "15kDhRAcpgsugmh6mQsTcCHdvbsuYncEEV",
+///    "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"
+///  ],
+///  "tags": [
+///    "/example/",
+///    "Example Pool"
+///  ],
+///  "link": "https://example.com"
+/// }
+/// ]
+/// ```
+///
+/// This format is used and produced by e.g. https://github.com/bitcoin-data/mining-pools
+/// and https://github.com/bitcoin-data/mining-pools/tree/generated
+pub fn parse_json(json: &str) -> Vec<Pool> {
+    serde_json::from_str(json).unwrap()
```

This should return an error when parsing the JSON fails, rather than `panicing`.

---

Done